renderlib/gesture/gesture.h

```diff
@@ -0,0 +1,388 @@
+    int lineLoopBegin = kInvalidVertexIndex;
+    std::vector<VertsCode> verts;
+    std::vector<CommandRange> commands[kNumCommandsLists];
+    std::unique_ptr<GLGuiShader> shader;
+
+    // A texture atlas for GUI elements
+    // TODO: use bindless textures
+    uint32_t glTextureId = 0;
+
+    // remember selection code to reuse while dragging
+    uint32_t m_retainedSelectionCode = SelectionBuffer::k_noSelectionCode;
+
+    // Empty the commands/verts buffers, typically done after drawing the GUI.
+    void clearCommands()
+    {
+      lineLoopBegin = kInvalidVertexIndex;
+      currentCommand = nullptr;
+
+      verts.clear();
+      for (int i = 0; i < kNumCommandsLists; ++i) {
+        commands[i].clear();
+      }
+    }
+
+    // Add a draw command. There are multiple command sequences you can add to so
+    // to handle 3d objects, 2d overlay, etc...
+    // Commands are added *before* defining their geometry
+    inline void addCommand(Command c, CommandSequence index = CommandSequence::k3dStacked)
+    {
+      // Terminate any previous command by recording the index of the last vertex
+      // before starting a new command
+      for (int i = 0; i < kNumCommandsLists; ++i) {
+        if (!commands[i].empty()) {
+          if (commands[i].back().end == kInvalidVertexIndex) {
+            commands[i].back().end = verts.size();
+          }
+        }
+      }
+
+      int begin = verts.size();
+      int end = kInvalidVertexIndex; //< leave the command buffer open to any new vertex
+
+      commands[static_cast<int>(index)].push_back({ c, begin, end });
+      currentCommand = &commands[static_cast<int>(index)].back();
+    }
+
+    // Any of GL_POINTS, GL_LINES, GL_TRIANGLES, etc...
+    // TODO: abstract the command type into an api-independent enum
+    inline void addCommand(GLenum command, CommandSequence index = CommandSequence::k3dStacked)
+    {
+      addCommand(Command(command), index);
+    }
+
+    // Add one vertex to the current command geometry
+    inline void addVert(const VertsCode& v)
+    {
+      assert(currentCommand != nullptr);
+      verts.push_back(v);
+    }
+
+    // Add two points to form a line, or the first line of a line loop.
+    // To extend the line to form a polyline (open or close) use extLine.
+    // It must be used with a GL_LINES command only.
+    inline void addLine(const VertsCode& v0, const VertsCode& v1)
+    {
+      assert(currentCommand != nullptr);
+      assert(currentCommand->command.command == GL_LINES);
+
+      // Mark the spot for the first line vertex so that we can close a line
+      // loop if we want to.
+      lineLoopBegin = verts.size();
+      verts.push_back(v0);
+      verts.push_back(v1);
+    }
+
+    enum class LoopEntry : int
+    {
+      kContinue = 0,
+      kClose = 1
+    };
+    // Create a polyline (open or close) by extending a line created with
+    // addLine.
+    inline void extLine(const VertsCode& v, LoopEntry loop = LoopEntry::kContinue)
```

Comment:
Nit: Maybe worth expanding this to be `extendLine`, since `ext` is a bit ambiguous